work-pool/src/WorkPool.hs

```diff
@@ -0,0 +1,119 @@
+module WorkPool
+  ( Config(..)
+  , Done
+  , Pool
+  , Source
+  , pushJob
+  , readJobCount
+  , runPool
+  , runSource
+  )
+where
+
+import Data.Set (Set)
+import MPrelude
+import Prelude (succ)
+
+import qualified Data.Set       as Set
+import qualified UnliftIO.Async as UnliftIO
+import qualified UnliftIO.STM   as UnliftIO
+
+-- | Worker pool configuration
+data Config m a = Config
+  { produceJobs :: Pool a -> m ()
+  -- ^ function called to produce work, use `pushJob` to create workable jobs.
+  -- once this function returns and all jobs are worked off the workers exit.
+  , queueSize   :: Natural
+  -- ^ maximum size of the jobs queued
+  , workerCount :: Natural
+  -- ^ number of workers to boot
+  , workerRun   :: Natural -> Source a -> m Done
+  -- ^ function called when a worker is booted, argument is the worker index,
+  -- and a source to be drained with `runSource`
```

Comment:
__suggestion:__ If nothing yet uses the index, I would probably remove it from the interface and add it when it is needed.